app/controllers/declaration/AdditionalInformationController.scala

```diff
@@ -39,64 +42,117 @@ class AdditionalInformationController @Inject()(
+      .fold(
+        formWithErrors =>
+          Future.successful(BadRequest(additional_information(itemId, appConfig, formWithErrors, cachedData))),
+        updatedCache =>
+          updateCacheModelAndRedirect(
+            itemId,
+            updatedCache,
+            controllers.declaration.routes.AdditionalInformationController.displayPage(itemId)
+        )
+      )
 
-        case Some(Remove(ids)) => {
-          val updatedCache = remove(ids.headOption, cache.items)
+  private def handleSaveAndContinue(
+    itemId: String,
+    boundForm: Form[AdditionalInformation],
+    cachedData: Seq[AdditionalInformation]
+  )(implicit request: JourneyRequest[_]): Future[Result] =
+    MultipleItemsHelper
+      .saveAndContinue(boundForm, cachedData, true, elementLimit)
+      .fold(
+        formWithErrors =>
+          Future.successful(BadRequest(additional_information(itemId, appConfig, formWithErrors, cachedData))),
+        updatedCache =>
+          if (updatedCache != cachedData)
+            updateCacheModelAndRedirect(
+              itemId,
+              updatedCache,
+              controllers.declaration.routes.DocumentsProducedController.displayPage(itemId)
+            )
+          else
+            Future.successful(Redirect(controllers.declaration.routes.DocumentsProducedController.displayPage(itemId)))
+      )
 
-          customsCacheService
-            .cache[AdditionalInformationData](goodsItemCacheId, formId, AdditionalInformationData(updatedCache))
-            .map(_ => Redirect(controllers.declaration.routes.AdditionalInformationController.displayForm()))
-        }
+  private def handleRemove(
+    itemId: String,
+    ids: Seq[String],
+    boundForm: Form[AdditionalInformation],
+    items: Seq[AdditionalInformation]
+  )(implicit request: JourneyRequest[_]): Future[Result] = {
+    val updatedCache = MultipleItemsHelper.remove(ids.headOption, items)
+    updateCacheModelAndRedirect(
+      itemId,
+      updatedCache,
+      controllers.declaration.routes.AdditionalInformationController.displayPage(itemId)
+    )
+  }
 
-        case Some(SaveAndContinue) =>
-          saveAndContinue(boundForm, cache.items, true, elementLimit).fold(
-            formWithErrors =>
-              Future.successful(BadRequest(additional_information(appConfig, formWithErrors, cache.items))),
-            updatedCache =>
-              if (updatedCache != cache.items)
-                customsCacheService
-                  .cache[AdditionalInformationData](goodsItemCacheId, formId, AdditionalInformationData(updatedCache))
-                  .map(_ => Redirect(controllers.declaration.routes.DocumentsProducedController.displayForm()))
-              else
-                Future.successful(Redirect(controllers.declaration.routes.DocumentsProducedController.displayForm()))
-          )
+  private def updateCacheModelAndRedirect(itemId: String, updatedCache: Seq[AdditionalInformation], redirectCall: Call)(
```

Comment:
Just my personal view but generally having `and` in a method name suggests it's doing a little too much.

I think it would be nice if we could break this up into an `updateCache` method which just does the `for` comprehension and yields `Future[Unit]` which we then just map to a `Redirect`

Reply:
I don't mind this as it removes a few more lines of code and the actual code in the function is not that long / complex.